src/role-helpers.js

```diff
@@ -0,0 +1,90 @@
+import {elementRoles} from 'aria-query'
+import {debugDOM} from './query-helpers'
+
+const elementRoleList = buildElementRoleList(elementRoles)
+
+function getImplicitAriaRoles(currentNode) {
+  for (const {selector, roles} of elementRoleList) {
+    if (currentNode.matches(selector)) {
+      return [...roles]
+    }
+  }
+
+  return []
+}
+
+function buildElementRoleList(elementRolesMap) {
+  function makeElementSelector({name, attributes = []}) {
+    return `${name}${attributes
+      .map(({name: attributeName, value}) => `[${attributeName}=${value}]`)
+      .join('')}`
+  }
+
+  function getSelectorSpecificity({attributes = []}) {
+    return attributes.length
+  }
+
+  function bySelectorSpecificity(
+    {specificity: leftSpecificity},
+    {specificity: rightSpecificity},
+  ) {
+    return rightSpecificity - leftSpecificity
+  }
+
+  let result = []
+
+  for (const [element, roles] of elementRolesMap.entries()) {
+    result = [
+      ...result,
+      {
+        selector: makeElementSelector(element),
+        roles: Array.from(roles),
+        specificity: getSelectorSpecificity(element),
+      },
+    ]
+  }
+
+  return result.sort(bySelectorSpecificity)
+}
+
+function getRoles(container) {
+  function flattenDOM(node) {
+    return [
+      node,
+      ...Array.from(node.children).reduce(
+        (acc, child) => [...acc, ...flattenDOM(child)],
+        [],
+      ),
+    ]
+  }
+
+  return flattenDOM(container).reduce((acc, node) => {
+    const roles = getImplicitAriaRoles(node)
+
+    return roles.reduce(
+      (rolesAcc, role) =>
+        Array.isArray(rolesAcc[role])
+          ? {...rolesAcc, [role]: [...rolesAcc[role], node]}
+          : {...rolesAcc, [role]: [node]},
+      acc,
+    )
+  }, {})
+}
+
+function logRoles(container) {
+  const roles = getRoles(container)
+
+  return Object.entries(roles)
+    .map(([role, elements]) => {
+      const numDelimeters = 42 - role.length // 42 is arbitrary
```

Comment:
As far as I know there aren't any aria-roles greater than 42 chars, but I would really hate for a formatting error to break someone's test. Can we be conservative here and do a Math.min(0, numDelimiters) or move the delimiter characters to a fixed length line?

Reply:
That's a good point.  I think it's very unlikely there will ever be a role greater than 42 characters long... there's also no real downside to protecting against that.  Maybe I was being too clever with embedding the name in the delimiters.  It read easier (to my eyes) with just a single heading/spacer element, but it probably makes more sense to just implement it like Kent originally had it and avoid any confusion, extra code, or bugs (however unlikely). 

I'll make that change this evening unless I hear an objection.

Reply:
Done [593c33f](https://github.com/testing-library/dom-testing-library/pull/285/commits/593c33f987bba473a3c0c89f446e8a1af604d7fd)

Check the snapshot for output format.  

I would have like to indent the elements themselves to further distinguish them from the headings, but unfortnately I couldn't find a decent way to do that with the way debugDOM->prettyDOM->prettyFormat works.  They inject their own newlines into the output, so if I try to indent it I can only indent the first line.  I'd need to split the output of debugDOM up by lines and indent them individually... which seemed messy for a minor visual improvement.  

I also purposefully left out the space between the delimeter bar ('-----') and the role name label.  With the space it looks a lot louder visually. 

     --------------------------------------------------
     navigation:
    <nav
        data-testid="a-nav"
    />

    --------------------------------------------------
    heading:

    <h1
      data-testid="a-h1"
    />

vs

     --------------------------------------------------

     navigation:
    <nav
        data-testid="a-nav"
    />

    --------------------------------------------------

    heading:

    <h1
      data-testid="a-h1"
    />